### DISCUSSION TITLE
.msi bootstraper for no UI mode

### DISCUSSION THREAD

**Discussion by akaJuliaan** [oleg-shilo/wixsharp#1520] [Q&A]
I would like to install my service, which is created in .net 8, with the help of a group policy. It is very likely that .net 8 isn't installed on the machines. Therefore I would like to install it automatically before my service get's installed. As I want to install my service with the help of a group policy, my installer needs to be a msi, because gpo's only support the installation of .msi files. Unfortunately I can't easily use a Simplified Bootstraper as the InstallUISequence is not fired in the no UI mode. But it seems that all the "normal" bootstrapers produce a .exe file which calls the msi's.

Is there any way to get a .msi bootstraper or to ensure that .net 8 will be installed before the installation of the service starts?

**Comment by oleg-shilo** [comment]
Yes, the SimplifiedBootstrapper is targeting UI-based installs. You are right.

WiX or NSIS bootstrapper seems to be the right tool for that. And you are right again. The bootstrapper is nothing else but an exe that embeds an MSIs and simply invokes them at runtime. You can even easily create a simple custom console-based exe that does exactly that with ~40 likes of code. "External_UI/ConsoleSetup" sample is the one that shows the technique.

However if you have such a constraint that you need to have a single msi then it is more complicated but doable.

1. Ultra conservative approach
    Have a check for the presence of .NETCore on you target system. If not detected then report (log/messagebox) the problem, advise how to fix (e.g. "Please install .NET8") and abort the installation.
    Code samples: <Wix# Samples>\LaunchConditions\LaunchConditions.csproj
          <Wix# Samples>\Managed Setup\setup.cs (does it in the MSI OnLoad event)

2. Conservative proactive approach
    The same as in point 1 but tell user that you will now initiate the installation of the .NET8 (download, and start the .net install process). Exit with the prompt to reattempt the service installation when .NET install is completed.
     
3. Reverse-bootstrapper
   Instead of exe build your msi that orcestrates the whole installation. The same as pinpoint 1 but have your builds script embedded in your msi:
    ```c#
    new ManagedProject(..., 
        new Binary(new Id("buildscript"), Assembly.GetExecutingAssembly().Location),
        . . .        
    ```
    In your `project.Load += msi_Load` even extract "buildscript" executable in the temp folder and start it. When starting pass the following information to the executable as arguments:
    mode - indicator that we are not building the msi but installing it
    msiLocation - path to the msi that has started this installation process. ou can get the location from the event arg `e.MsiFile` or the session object (`session.Property("OriginalDatabase");`)
    parentProcess - process Id that is forking the child buildscript app process  
    Now you can exit.
    Then the child process will detect (from mode arg) that it needs to bootstrap the install.
    Then it will wait until the parent gracefully exit (you have the id for that).
    Next it will install .NET8
    Next it will start the product installation again (`msiexec /i msiLocation`)
    ........
    Technically this approach is very simple. I even have some samples for that. It is the inter-process implications that you need to consider and decide if you want to go this way.

**Reply by akaJuliaan** [reply]
Thank you very much for your reply!
